_source/_posts/2019-12-20-five-tools-improve-java.adoc

```diff
@@ -0,0 +1,345 @@
+
+PMD doesn't detect all the issues with this code though, the observant reader may have noticed a few bigger problems. To detect those, we can use SpotBugs.
+
+== 3. Conduct ByteCode Analysis with SpotBugs and Find Security Bugs
+
+image:{% asset_path 'blog/five-tools-improve-java/spotbugs-logo.png' %}[alt=Spotbugs logo ...,,align=center]
+
+SpotBugs checks byte code, whereas PMD scans source files, this means anything the compiler throws away (unused imports or example) would not be reported by SpotBugs. The SpotBugs project is an updated version of FindBugs, many items in the documentation still reference "FindBugs." For those of you still using FindBugs, https://spotbugs.readthedocs.io/en/stable/migration.html[updating to SpotBugs] is trivial.
+
+SpotBugs also has a few plugins, my favorite being "Find Security Bugs", and as you might guess, it helps you detect security issues like weak hash functions, file/path traversals, untrusted inputs, https://find-sec-bugs.github.io/bugs.htm[and many more].
+
+To add SpotBugs (and Find Security Bugs) to a Maven project, add the following to your `pom.xml`:
+
+.pom.xml
+[source,xml]
+----
+<plugin>
+    <groupId>com.github.spotbugs</groupId>
+    <artifactId>spotbugs-maven-plugin</artifactId>
+    <version>3.1.12.2</version>
+    <configuration>
+        <effort>Max</effort>
+        <threshold>Low</threshold>
+        <failOnError>true</failOnError>
+        <plugins>
+            <plugin>
+                <groupId>com.h3xstream.findsecbugs</groupId>
+                <artifactId>findsecbugs-plugin</artifactId> <!--1-->
+                <version>1.10.1</version>
+            </plugin>
+        </plugins>
+    </configuration>
+    <executions>
+        <execution>
+            <id>scan</id>
+            <phase>verify</phase>
+            <goals>
+                <goal>check</goal>
+            </goals>
+        </execution>
+    </executions>
+</plugin>
+----
+<1> Find Security Bugs is a plugin to a plugin 🤯
+
+When we scan the same code as above using `mvn compile spotbugs:check`, typically you would just run `mvn verify`, however, we haven't fixed the PMD issues above yet.
+
+TIP: **NOTE:** you can skip PMD using the command line arg `-Dpmd.skip`, similarly with SpotBugs, `-Dspotbugs.skip`. For example: `mvn verify -Dpmd.skip -Dspotbugs.skip` would skip both.
+
+.FileUtility.java
+[source,java]
+----
+package com.okta.examples;
+
+import java.io.File;
+import java.io.FileWriter;;
+import java.io.IOException;
+import java.io.BufferedOutputStream;
+
+public class FileUtility {
+
+    private static void toFile_bad(String contents, File file) throws IOException { // <3>
+        FileWriter fileWriter = new FileWriter(file); // <1> <2>
+        fileWriter.write(contents);
+    }
+}
+----
+<1> Default encoding used, use `UTF8` or other Charset.
+<2> Failed to close FileWriter, consider using a try-with-resources block.
+<3> Unused private method - dead code, consider deleting.
```

Comment:
A cool tip I learned recently is you can use `<.>` in the code and in the numbers if they're sequential.  https://twitter.com/mrhaki/status/1204289426588946432